### DISCUSSION TITLE
Very large k values for ~200M vectors

### DISCUSSION THREAD

**Discussion by mataney** [milvus-io/milvus#15306] [Q&A and General discussion] [answered]
Hi 👋 
Thank's for the great work!
I have a dataset of around 2M vectors of size 768 (BERT embeddings, but (a) I can lower the dimensionality, (b) for the sake of argument let's assume they are just random vectors) and I'd like to do approximate NN search on that dataset.
My requirement is that my k value is very large - around 50,000.

IIUC max k is 1024/2048, depending on my GPU.
But what if I will shard my index to 25-50 different shards? Can this do the trick?
What do you think it will do to the latency?
Is there a tutorial that shows something remotely like this?

Thank you for your help

**Comment by xiaofan-luan** [comment]
Hi mataney, which version you are using?thought if are using GPU then its 1.1?

**Reply by mataney** [reply]
`pymilvus 2.0.0rc9`
Isn't is weird I got the `is too large!` exception?

Maybe because I'm using the python wrapper?

**Reply by yhmo** [reply]
pymilvus 2.0.0rc9 is for Milvus 2.0 Pre-GA version, the topk limit is 16384

Milvus 1.1 requires pymilvus 1.1.2:
`pip3 install pymilvus==1.1.2`

**Comment by yhmo** [comment, marked as answer]
In Milvus 1.x, version 1.1.1 allow user to input a large topk up to 1M. The release note is here: https://github.com/milvus-io/milvus/releases/tag/v1.1.1

For 1.1.1 GPU version, if a user inputs a topk larger than 16384, the Milvus will use CPU to compute. If you want use GPU with large topk(>16384), yes you can shard it by yourself. For example, deploy 5 milvus standalone servers, each server manages 400,000 vectors, send a search request to 5 servers, topk=10000, then merge the results returned by 5 servers.

For Milvus 2.0, the topk limit is 16384. The 2.0 doesn't support GPU yet. In our plan, GPU support is a feature of 2.1.

**Reply by mataney** [reply]
Thanks for you reply!
Sounds great. 

Can you please point me to the tutorial that shows how to shard effectively?

**Reply by yhmo** [reply]
> Thanks for you reply! Sounds great.
> 
> Can you please point me to the tutorial that shows how to shard effectively?

The milvus 1.1 has a middleware to do sharding: https://github.com/milvus-io/milvus/tree/1.1/shards
But I don't recommend you use it. Because it is mainly for large dataset.
If you don't care how milvus manage the data, just want high performance, I suggest your use this solution:
1. start 5 milvus servers: S_1, S_2, S_3, S_4, S_5, modify the server_config.yaml to order them use GPU
```
gpu:
  enable: true
  cache.enable: true
  cache_size: 2GB
  gpu_search_threshold: 1
  search_devices:
    - gpu0
```

2. call `create_collection()` to each server, with the same collection name "my_collection"
3. call `insert()` to each server, insert 400k vectors into "my_collection" collection
4. call `create_index()` to each server, create IVF index(IVFFLAT or IVFSQ8)
5. call `load_collection()` to each server
6. call `search()` to each server, with the same search parameters, topk=10000
7. get results from 5 servers, each result has 10000 id-distance pairs, combine them into 50000 id-distance pairs. You may arrange the pairs by distance value.
